File: Sort/InsertionSortList.java
/*
Sort a linked list using insertion sort.
*/

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode insertionSortList(ListNode head) {
        if(head == null || head.next == null) return head;
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode pre = head;
        while(pre.next != null){
            ListNode temp = pre.next;
            ListNode cur = pre.next;
            ListNode p = dummy;
            while(p.next != cur){
                ListNode next = p.next;
                if(cur.val < next.val){
                    pre.next = cur.next;
                    p.next = cur;
                    cur.next = next;
                    break;
                }
                else p = next;
            }
            if(pre.next == temp) pre = pre.next;
        }
        return dummy.next;
    }
}

File: Sort/InsertionSortList.js
/*
Sort a linked list using insertion sort.
*/

/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var insertionSortList = function(head) {
    if(head === null || head.next === null) return head;
    var dummy = new ListNode(0);
    dummy.next = head;
    var pre = head;
    while(pre.next !== null){
        var cur = pre.next;
        var temp = pre.next
        var p = dummy;
        while(p.next != cur){
            var next = p.next;
            if(next.val > cur.val){
                pre.next = cur.next;
                p.next = cur;
                cur.next = next;
            }
            else p = p.next;
        }
        if(pre.next == temp) pre = pre.next;
    }
    return dummy.next;
};